internal/querywatcher/query_watcher.go

```diff
@@ -163,35 +187,56 @@ func (w *QueryWatcher) updateQueryCache(query *DSQLQuery, event dstore.WatchEven
 	}
 }
 
-// notifyClients notifies all clients watching a query about the new result.
-func (w *QueryWatcher) notifyClients(query *DSQLQuery, clients *sync.Map, queryResult *[]dstore.DSQLQueryResultRow) {
+// notifyClient notifies all clients watching a query about the new result.
+func (w *QWatchManager) notifyClient(query *DSQLQuery, clientFD int, queryResult *[]dstore.DSQLQueryResultRow) {
 	encodedResult := clientio.Encode(CreatePushResponse(query, queryResult), false)
-	clients.Range(func(clientKey, _ interface{}) bool {
-		clientFD := clientKey.(int)
-		_, err := syscall.Write(clientFD, encodedResult)
-		if err != nil {
-			w.removeWatcher(query, clientFD)
-		}
-		return true
-	})
+	// Send the result to the client
+	_, err := syscall.Write(clientFD, encodedResult)
+	if err != nil {
+		w.removeWatcher(query, clientFD)
+	}
 }
 
 // serveAdhocQueries listens for adhoc queries, executes them, and sends the result back to the client.
-func (w *QueryWatcher) serveAdhocQueries(ctx context.Context) {
+func (w *QWatchManager) serveAdhocQueries(ctx context.Context) {
 	for {
 		select {
 		case query := <-AdhocQueryChan:
 			result, err := w.runQuery(&query.Query)
-			query.ResponseChan <- AdhocQueryResult{Result: result, Err: err}
+			query.ResponseChan <- AdhocQueryResult{
+				Result:      result,
+				Fingerprint: generateQueryFingerprint(&query.Query),
+				Err:         err,
+			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+// generateQueryFingerprint used to creating a common fingerprint for similar queries
+func generateQueryFingerprint(query *DSQLQuery) string {
+	hash := sha256.Sum256([]byte(query.KeyRegex))
```

Comment:
I don't think sha256 is a good choice here since we don't care about cryptographic security. We want something fast, maybe something like FarmHash?

Reply:
We can also look into `MurmurHash` currently already being used in project as part of `bloom` implementation.

Reply:
I think FarmHash is a newer and more efficient implementation

Reply:
Used Farmhash to generate the fingerprint.